eureka2-core/src/main/java/com/netflix/eureka2/health/AbstractHealthStatusProvider.java

```diff
@@ -0,0 +1,39 @@
+package com.netflix.eureka2.health;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.netflix.eureka2.registry.instance.InstanceInfo.Status;
+import rx.Observable;
+import rx.subjects.BehaviorSubject;
+
+/**
+ * @author Tomasz Bak
+ */
+public abstract class AbstractHealthStatusProvider<SUBSYSTEM> implements HealthStatusProvider<SUBSYSTEM> {
+
+    private final AtomicReference<Status> currentStatus;
+    private final SubsystemDescriptor<SUBSYSTEM> descriptor;
+
+    private final BehaviorSubject<HealthStatusUpdate<SUBSYSTEM>> healthSubject = BehaviorSubject.create();
+
+    protected AbstractHealthStatusProvider(Status initialState, SubsystemDescriptor<SUBSYSTEM> descriptor) {
+        this.currentStatus = new AtomicReference<>();
+        this.descriptor = descriptor;
+        moveHealthTo(initialState);
+    }
+
+    @Override
+    public Observable<HealthStatusUpdate<SUBSYSTEM>> healthStatus() {
+        return healthSubject;
+    }
+
+    public boolean moveHealthTo(Status newStatus) {
+        if (currentStatus.getAndSet(newStatus) == newStatus) {
+            return false;
+        }
+        healthSubject.onNext(new HealthStatusUpdate<SUBSYSTEM>(newStatus, descriptor));
```

Comment:
Is concurrency control important here? There is still no concurrency guard against two threads that execute in the following order:
Thread1               Thread2
S1 -> S2              
                            S2 -> S3
                            onNext S3
onNext S2